docs/proposals/kubelet-raw-metrics-api.md

```diff
@@ -0,0 +1,120 @@
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubelet raw metrics API
+
+Kubelet currently exposes raw container metrics through the `/stats/` endpoint
+that serves raw container stats. However, this endpoint serves individual
+container stats, and applications like heapster, which aggregates metrics across
+the cluster, must repeatedly query for each container. The high QPS combined
+with the potential data size of raw stats puts unnecessary load on the system
+that could be avoided with an aggregate API.
+
+## Requirements:
+- Provide machine level metrics, all pod metrics (in single request), specific
+  pod metrics
+- Ability to authenticate machine & pod metrics independently from each other
+- Eventually support multiple kinds of metrics (e.g. derived/aggregate type)
+- Follow existing API conventions, fully compatible types able to eventually be
+  served by apiserver library
+
+## Design
+
+Kubelet will start providing a versioned API, following convensions of the API
+server. The metrics API will be its own [API group](api-group.md), and may
+eventually be reused at the cluster-level (outside the scope of this proposal).
+
+### Endpoints
+
+All endpoints are GET endpoints, rooted at `/apis/metrics/v1alpha1/`
+
+- `/` - discovery endpoint; type resource list
+- `/nodes` - host metrics; type `[]metrics.Node`
+  - `/nodes/localhost` - When served by kubelet, the only node provided is
+    `localhost`; type metrics.Node
+- `/pods` - All pod metrics across all namespaces; type `[]metrics.Pod`
+- `/namespaces/{namespace}/pods` - All pod metrics within namespace; type
+  `[]metrics.Pod`
+- `/namespaces/{namespace}/pods/{pod}` - metrics for specific pod; type
+  `metrics.Pod`
+- Unsupported paths return status not found (404)
+  - `/namespaces/`
+  - `/namespaces/{namespace}`
+
+Additionally, all endpoints (except root discovery endpoint) support the
+following optional query parameters:
+
+- `start` - start time to return metrics from; type json encoded `time.Time`
+- `end` - end time to return metrics to; type json encoded `time.Time`
+- `count` - maximum number of stats to return in each ContainerMetrics instance;
+  type int
+
+Future metrics types, such as derived metrics, will add a prefix to relevant
+supported endpoints. E.g.
+
+- `/derivedNodes(/localhost)`
+- `/derivedPods`
+- `/namespaces/{namespace}/derivedPods(/{pod})`
+
+### Schema
+
+Types are colocated with other API groups in `/pkg/apis/metrics`, and follow api
+groups conventions there. Types (pseudo-go) are:
+
+```go
+type Node struct {
+  ObjectMeta, TypeMeta
+  Machine ContainerMetrics
+  SystemContainers []ContainerMetrics
+}
+type Pod struct {
+  ObjectMeta, TypeMeta
+  Containers []ContainerMetrics
+}
+
+// ContainerMetrics is a k8s wrapper around cAdvisor metrics
+type ContainerMetrics struct {
+  ObjectMeta, TypeMeta
+  Info cadvisorv2.ContainerInfo
```

Comment:
I'm sold. We'll mirror the cadvisor v2 APIs in the versioned metrics API, and also add ObjectMeta & TypeMeta to each object.